src/views/PianoRoll/FrameListGui.lua

```diff
@@ -0,0 +1,297 @@
+        text = pianoRoll.name
+    })
+
+    draw:text(grid_rect(col0, row2, col1 - col0, 1), "start", "Frame")
+    draw:text(grid_rect(col1, row2, col6 - col1, 1), "start", "Joystick")
+    draw:text(grid_rect(col6, row2, col_1 - col6, 1), "start", "Buttons")
+
+    local rect = grid_rect(0, row3 + 0.35, 0.333, 0.333)
+    for i, v in ipairs(Buttons) do
+        rect.x = buttonDrawData[i].x
+        draw:small_text(rect, "center", v.text)
+    end
+end
+
+local function DrawColorCodes()
+    local offset = 0
+
+    local width = small * 1.1
+    local rect = grid_rect(col6 + offset, row4, width, 0.5)
+    local height = rect.height * NumDisplayFrames()
+
+    local i = 1
+    local colorIndex = 1
+    local buttonDrawData = {}
+
+    local function DrawNext(amount)
+        for k = 0, amount - 1, 1 do
+            buttonDrawData[i] = {x = rect.x + k * rect.width, colorIndex = colorIndex}
+            i = i + 1
+        end
+        BreitbandGraphics.fill_rectangle(
+            {x = rect.x, y = rect.y, width = rect.width * amount, height = height},
+            buttonColors[colorIndex].background
+        )
+        rect.x = rect.x + rect.width * amount
+        colorIndex = colorIndex + 1
+    end
+
+    DrawNext(1)
+    DrawNext(1)
+    DrawNext(1)
+    DrawNext(1)
+    DrawNext(4)
+    DrawNext(2)
+    DrawNext(4)
+
+    return buttonDrawData
+end
+
+local placing = false
+local function PlaceAndUnplaceButtons(frameRect, currentInput, pressed, buttonDrawData)
+    local mouseX = uguiInputContext.mouse_position.x
+    local relativeY = uguiInputContext.mouse_position.y - frameRect.y
+    local inRange = mouseX >= frameRect.x and mouseX <= frameRect.x + frameRect.width and relativeY >= 0
+    local frameIndex = math.floor(relativeY / frameRect.height)
+    local hoveringGlobalTimer = frameIndex + scrollOffset + PianoRollContext.current.startGT
+    local frame = PianoRollContext.current.frames[hoveringGlobalTimer]
+    local anyChange = false
+    inRange = inRange and frameIndex <= PianoRollContext.maxDisplayedFrames
+    UpdateScroll(inRange and uguiInputContext.wheel or 0)
+
+    if inRange and frame ~= nil then
+        for buttonIndex, v in ipairs(Buttons) do
+            local inRangeX = mouseX >= buttonDrawData[buttonIndex].x and mouseX < (buttonDrawData[buttonIndex + 1] or {x=9999999}).x
+            if pressed.leftclick and inRangeX then
+                placing = not frame.joy[v.input]
+                frame.joy[v.input] = placing
+                anyChange = true
+            elseif currentInput.leftclick and inRangeX then
+                anyChange = frame.joy[v.input] ~= placing
+                frame.joy[v.input] = placing
+            end
+        end
+    end
+    return anyChange
+end
+
+local function DrawFramesGui(pianoRoll, draw, buttonDrawData)
+
+    local currentInput = input.get()
+    local pressed = input.diff(currentInput, lastInput)
```

Comment:
Hmm, maybe it'd be best if the ugui interface would simply publicly expose some of the functionality it uses internally.  
It could be something as simple as:
```Lua
ugui.inputs = {
    is_mouse_just_down = ugui.internal.is_mouse_just_down,
    is_mouse_wheel_up = ugui.internal.is_mouse_wheel_up,
    ...
}
```

Reply:
While that would be more formally sane, it should be fine to call the internal API since it's not necessarily a particularly volatile part of the library